src/inmanta/agent/agent.py

```diff
@@ -1332,63 +1617,98 @@ async def on_disconnect(self) -> None:
+        invalid_resource_types: set[str] = set()
+        for resource_type in set(resource_types):
+
+            cached_spec: Optional[ResourceInstallSpec] = self._last_loaded.get((resource_type, version))
+            if cached_spec:
+                LOGGER.debug("Code already present for %s version=%d", resource_type, version)
+                resource_install_specs.append(cached_spec)
+                continue
+            result: protocol.Result = await self._client.get_source_code(environment, version, resource_type)
+            if result.code == 200 and result.result is not None:
+                sync_client = SyncClient(client=self._client, ioloop=self._io_loop)
+                LOGGER.debug("Installing handler %s version=%d", resource_type, version)
+                requirements: set[str] = set()
+                sources: list["ModuleSource"] = []
+                # Encapsulate source code details in ``ModuleSource`` objects
+                for source in result.result["data"]:
+                    sources.append(
+                        ModuleSource(
+                            name=source["module_name"],
+                            is_byte_code=source["is_byte_code"],
+                            hash_value=source["hash"],
+                            _client=sync_client,
+                        )
+                    )
+                    requirements.update(source["requirements"])
+
+                if pip_config is None:
+                    try:
+                        pip_config = await self._get_pip_config(environment, version)
+                    except Exception:
+                        invalid_resource_types.add(resource_type)
+                        continue
+
+                resource_install_specs.append(
+                    ResourceInstallSpec(resource_type, version, pip_config, list(requirements), sources)
+                )
+            else:
+                invalid_resource_types.add(resource_type)
+
+        return resource_install_specs, invalid_resource_types
+
+    async def ensure_code(self, code: Sequence[ResourceInstallSpec]) -> Set[str]:
+        """Ensure that the code for the given environment and version is loaded"""
+        failed_to_load: set[str] = set()
+        # breakpoint()
+        for resource_install_spec in code:
             # only one logical thread can load a particular resource type at any time
-            async with self._resource_loader_lock.get(rt):
+            async with self._resource_loader_lock.get(resource_install_spec.resource_type):
                 # stop if the last successful load was this one
                 # The combination of the lock and this check causes the reloads to naturally 'batch up'
-                if self._last_loaded[rt] == version:
-                    LOGGER.debug("Code already present for %s version=%d", rt, version)
+                if self._last_loaded.get((resource_install_spec.resource_type, resource_install_spec.model_version)):
```

Comment:
I don't think this holds anymore: `_last_loaded` doesn't really represent "last loaded" anymore. Just "loaded at some point. Either we have to change it back to mean "last loaded" or we have to be smarter about it.

Reply:
Added back a dict to keep track of the actual last loaded version + fixed a bug where the try..catch surrounding the installation below was executing outside of the `_resource_loader_lock`